File: src/types/git.ts
export type TGitChangeKind =
  | 'added'
  | 'modified'
  | 'deleted'
  | 'renamed'
  | 'typechange'
  | 'untracked'
  | 'conflicted';

export type TGitDiffMode = 'worktree' | 'staged';

export type TGitBranchKind = 'local' | 'remote';

export type TGitPullRequestProvider =
  | 'github'
  | 'gitlab'
  | 'gitea'
  | 'bitbucket'
  | 'unknown';

export interface IGitCommitSummaryPayload {
  id: string;
  shortId: string;
  summary: string;
  authorName: string;
  authoredAt: string;
}

export interface IGitCommitHistoryRequest {
  repositoryRootPath: string;
  offset?: number;
  limit?: number;
}

export interface IGitCommitHistoryPayload {
  entries: IGitCommitSummaryPayload[];
  hasMore: boolean;
  nextOffset: number | null;
}

export interface IGitFileStatusPayload {
  path: string;
  relativePath: string;
  fileName: string;
  previousPath: string | null;
  previousRelativePath: string | null;
  indexStatus: TGitChangeKind | null;
  worktreeStatus: TGitChangeKind | null;
  isConflicted: boolean;
  isUntracked: boolean;
}

export interface IGitBranchPayload {
  name: string;
  shorthand: string;
  kind: TGitBranchKind;
  upstreamName: string | null;
  isCurrent: boolean;
  isHead: boolean;
  ahead: number;
  behind: number;
  lastCommit: IGitCommitSummaryPayload | null;
}

export interface IGitBranchListPayload {
  branches: IGitBranchPayload[];
}

export interface IGitBranchCheckoutRequest {
  repositoryRootPath: string;
  branchName: string;
}

export interface IGitBranchCreateRequest {
  repositoryRootPath: string;
  branchName: string;
  checkout: boolean;
}

export interface IGitRepositoryStatusPayload {
  available: boolean;
  message: string | null;
  repositoryRootPath: string | null;
  repositoryName: string | null;
  gitDirPath: string | null;
  headBranchName: string | null;
  headShortName: string | null;
  headShortOid: string | null;
  isDetached: boolean;
  isClean: boolean;
  ahead: number;
  behind: number;
  stagedCount: number;
  unstagedCount: number;
  untrackedCount: number;
  conflictedCount: number;
  files: IGitFileStatusPayload[];
  lastCommit: IGitCommitSummaryPayload | null;
}

export interface IGitRepositoryRootRequest {
  repositoryRootPath: string;
}

export interface IGitFileBaselinePayload {
  available: boolean;
  message: string | null;
  repositoryRootPath: string | null;
  filePath: string;
  relativePath: string | null;
  isTracked: boolean;
  content: string | null;
}

export interface IGitStashEntryPayload {
  index: number;
  stashId: string;
  summary: string;
  branchName: string | null;
  commitShortId: string | null;
}

export interface IGitStashListPayload {
  entries: IGitStashEntryPayload[];
}

export interface IGitStashSaveRequest {
  repositoryRootPath: string;
  message: string | null;
  includeUntracked: boolean;
}

export interface IGitStashApplyRequest {
  repositoryRootPath: string;
  stashIndex: number;
  pop: boolean;
}

export interface IGitStashDropRequest {
  repositoryRootPath: string;
  stashIndex: number;
}

export interface IGitPullRequestSupportPayload {
  available: boolean;
  remoteName: string | null;
  provider: TGitPullRequestProvider;
  repositoryUrl: string | null;
  pullRequestsUrl: string | null;
  createPullRequestUrl: string | null;
}

export interface IGitDiffPreviewRequest {
  repositoryRootPath: string;
  path: string;
  mode: TGitDiffMode;
}

export interface IGitDiffPreviewPayload {
  id: string;
  repositoryRootPath: string;
  path: string;
  relativePath: string;
  title: string;
  mode: TGitDiffMode;
  originalContent: string;
  modifiedContent: string;
  isEmpty: boolean;
}

export interface IGitPathOperationRequest {
  repositoryRootPath: string;
  paths: string[];
}

export interface IGitCommitRequest {
  repositoryRootPath: string;
  message: string;
}

export interface IGitCommitResultPayload {
  status: IGitRepositoryStatusPayload;
  commit: IGitCommitSummaryPayload;
}

File: src-tauri/src/commands/git.rs
use super::{ decode_script_bytes, resolve_workspace_root };

use gix::bstr::ByteSlice;
use serde::{Deserialize, Serialize};
use std::{
    fs,
    path::{Component, Path, PathBuf},
};

mod cli;
mod branches;
mod diff;
mod history;
mod pull_request;
mod stash;
mod status;

#[cfg(test)]
mod tests;

pub use branches::{checkout_git_branch, create_git_branch, list_git_branches};
pub use diff::get_git_diff_preview;
pub use history::list_git_commit_history;
pub use pull_request::get_git_pull_request_support;
pub use stash::{apply_git_stash, drop_git_stash, list_git_stashes, save_git_stash};
pub use status::{
    commit_git_index, discard_git_paths, get_git_file_baseline, get_git_repository_status,
    init_git_repository, stage_git_paths, unstage_git_paths,
};

const GIT_DIFF_MODE_WORKTREE: &str = "worktree";
const GIT_DIFF_MODE_STAGED: &str = "staged";
const DEFAULT_GIT_HISTORY_LIMIT: usize = 20;
const MAX_GIT_HISTORY_LIMIT: usize = 200;

type Repository = gix::Repository;

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitSummaryPayload {
    id: String,
    short_id: String,
    summary: String,
    author_name: String,
    authored_at: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitHistoryRequest {
    repository_root_path: String,
    offset: Option<usize>,
    limit: Option<usize>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitHistoryPayload {
    entries: Vec<GitCommitSummaryPayload>,
    has_more: bool,
    next_offset: Option<usize>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchPayload {
    name: String,
    shorthand: String,
    kind: String,
    upstream_name: Option<String>,
    is_current: bool,
    is_head: bool,
    ahead: usize,
    behind: usize,
    last_commit: Option<GitCommitSummaryPayload>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchListPayload {
    branches: Vec<GitBranchPayload>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitRepositoryRootRequest {
    repository_root_path: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchCheckoutRequest {
    repository_root_path: String,
    branch_name: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchCreateRequest {
    repository_root_path: String,
    branch_name: String,
    checkout: bool,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitFileStatusPayload {
    path: String,
    relative_path: String,
    file_name: String,
    previous_path: Option<String>,
    previous_relative_path: Option<String>,
    index_status: Option<String>,
    worktree_status: Option<String>,
    is_conflicted: bool,
    is_untracked: bool,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitRepositoryStatusPayload {
    available: bool,
    message: Option<String>,
    repository_root_path: Option<String>,
    repository_name: Option<String>,
    git_dir_path: Option<String>,
    head_branch_name: Option<String>,
    head_short_name: Option<String>,
    head_short_oid: Option<String>,
    is_detached: bool,
    is_clean: bool,
    ahead: usize,
    behind: usize,
    staged_count: usize,
    unstaged_count: usize,
    untracked_count: usize,
    conflicted_count: usize,
    files: Vec<GitFileStatusPayload>,
    last_commit: Option<GitCommitSummaryPayload>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitFileBaselinePayload {
    available: bool,
    message: Option<String>,
    repository_root_path: Option<String>,
    file_path: String,
    relative_path: Option<String>,
    is_tracked: bool,
    content: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffPreviewRequest {
    repository_root_path: String,
    path: String,
    mode: String,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitDiffPreviewPayload {
    id: String,
    repository_root_path: String,
    path: String,
    relative_path: String,
    title: String,
    mode: String,
    original_content: String,
    modified_content: String,
    is_empty: bool,
}

struct GitDiffContentPair {
    original_content: String,
    modified_content: String,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitResultPayload {
    status: GitRepositoryStatusPayload,
    commit_id: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitRequest {
    repository_root_path: String,
    message: String,
    paths: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitPathOperationRequest {
    repository_root_path: String,
    paths: Vec<String>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitStashEntryPayload {
    index: usize,
    stash_id: String,
    summary: String,
    branch_name: Option<String>,
    commit_short_id: Option<String>,
    created_at: String,
    file_count: usize,
    additions: u32,
    deletions: u32,
    files: Vec<GitStashFilePayload>,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitStashFilePayload {
    relative_path: String,
    file_name: String,
    previous_relative_path: Option<String>,
    status: String,
    additions: u32,
    deletions: u32,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitStashListPayload {
    entries: Vec<GitStashEntryPayload>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStashSaveRequest {
    repository_root_path: String,
    message: Option<String>,
    include_untracked: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStashApplyRequest {
    repository_root_path: String,
    stash_index: usize,
    pop: bool,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStashDropRequest {
    repository_root_path: String,
    stash_index: usize,
}

#[derive(Debug, Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GitPullRequestSupportPayload {
    available: bool,
    remote_name: Option<String>,
    provider: String,
    repository_url: Option<String>,
    pull_requests_url: Option<String>,
    create_pull_request_url: Option<String>,
}

fn open_repository_from_root(root: &str) -> Result<Repository, String> {
    let root = normalize_path_for_git(Path::new(root));
    gix::open(&root).map_err(|error| format!("打开 Git 仓库失败：{error}"))
}

pub(super) fn resolve_repository_root(repository: &Repository) -> Result<PathBuf, String> {
    repository
        .workdir()
        .map(Path::to_path_buf)
        .ok_or_else(|| "当前 Git 仓库没有工作区。".to_string())
}

fn resolve_git_workspace_root(
    workspace_root_path: Option<String>,
) -> Result<PathBuf, String> {
    match workspace_root_path {
        Some(value) if !value.trim().is_empty() => {
            let path = normalize_path_for_git(Path::new(value.trim()));
            Ok(path)
        }
        _ => resolve_workspace_root(None),
    }
}

fn resolve_head_commit(repository: &Repository) -> Result<Option<gix::Commit<'_>>, String> {
    match repository.head_commit() {
        Ok(commit) => Ok(Some(commit)),
        Err(error) => {
            let err_str = format!("{error}");
            if err_str.contains("unborn")
                || err_str.contains("does not exist")
                || err_str.contains("not found")
            {
                Ok(None)
            } else {
                Err(format!("读取 Git HEAD 提交失败：{error}"))
            }
        }
    }
}



fn build_git_commit_summary(commit: &gix::Commit<'_>) -> GitCommitSummaryPayload {
    let authored_at = jiff::Timestamp::from_second(
        commit.time().unwrap_or_default().seconds as i64,
    )
    .unwrap_or_else(|_| jiff::Timestamp::now())
    .to_string();

    let summary = commit
        .message()
        .ok()
        .map(|m| m.summary().to_str_lossy().into_owned())
        .filter(|s| !s.is_empty())
        .unwrap_or_else(|| "无提交说明".to_string());

    GitCommitSummaryPayload {
        id: commit.id().to_string(),
        short_id: short_commit_id(commit.id().detach()),
        summary,
        author_name: commit
            .author()
            .map(|a| a.name.to_string())
            .unwrap_or_else(|_| "未知作者".to_string()),
        authored_at,
    }
}

fn short_commit_id(id: gix::ObjectId) -> String {
    id.to_string().chars().take(7).collect()
}

fn resolve_relative_path(repository_root: &Path, path: &Path) -> Result<PathBuf, String> {
    let path_candidate = if path.is_absolute() {
        path.to_path_buf()
    } else {
        repository_root.join(path)
    };
    let path_candidate = normalize_path_for_git(&path_candidate);

    path_candidate
        .strip_prefix(repository_root)
        .map(Path::to_path_buf)
        .map_err(|_| "目标文件超出当前 Git 仓库根目录。".to_string())
}

fn resolve_pathspecs(repository_root: &Path, paths: &[String]) -> Result<Vec<String>, String> {
    let mut pathspecs = Vec::new();

    for path in paths {
        if path.trim().is_empty() {
            continue;
        }

        let relative_path = resolve_relative_path(repository_root, Path::new(path))?;

        if relative_path
            .components()
            .any(|component| matches!(component, Component::ParentDir))
        {
            return Err(format!("Git 变更路径不合法：{path}"));
        }

        let pathspec = path_to_forward_slashes(&relative_path);
        if !pathspec.is_empty() {
            pathspecs.push(pathspec);
        }
    }

    Ok(pathspecs)
}

fn path_to_forward_slashes(path: &Path) -> String {
    path.to_string_lossy().replace('\\', "/")
}

#[cfg(windows)]
fn normalize_path_for_git(path: &Path) -> PathBuf {
    let value = path.to_string_lossy();

    if let Some(stripped) = value.strip_prefix(r"\\?\UNC\") {
        return PathBuf::from(format!(r"\\{stripped}"));
    }

    if let Some(stripped) = value.strip_prefix(r"\\?\") {
        return PathBuf::from(stripped.to_string());
    }

    if let Some(stripped) = value.strip_prefix("//?/UNC/") {
        return PathBuf::from(format!("//{stripped}").replace('/', r"\\"));
    }

    if let Some(stripped) = value.strip_prefix("//?/") {
        return PathBuf::from(stripped.replace('/', r"\\"));
    }

    path.to_path_buf()
}

#[cfg(not(windows))]
fn normalize_path_for_git(path: &Path) -> PathBuf {
    path.to_path_buf()
}